// @observablehq/notebook-kit/dist/src/vite/config.js
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { resolveNpmImport } from "../javascript/imports/npm.js";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
export function config() {
    return {
        base: "./",
        appType: "mpa", // return 404 for missing pages
        esbuild: {
            supported: {
                "top-level-await": true
            }
        },
        resolve: {
            alias: [
                {
                    find: /^(npm:.*)$/,
                    replacement: "$1",
                    customResolver: (source) => ({ id: resolveNpmImport(source), external: true })
                },
                {
                    find: /^jsr:(.*)$/,
                    replacement: "https://esm.sh/jsr/$1"
                },
                {
                    find: /^observable:(.*)$/,
                    replacement: resolve(__dirname, "../$1")
                }
            ]
        }
    };
}
